src/main/java/io/r2dbc/mssql/MssqlConnectionFactory.java

```diff
@@ -50,28 +52,68 @@
      * @throws IllegalArgumentException when {@link MssqlConnectionConfiguration} is {@code null}.
      */
     public MssqlConnectionFactory(MssqlConnectionConfiguration configuration) {
-        this(Mono.defer(() -> {
-            Assert.requireNonNull(configuration, "configuration must not be null");
-
-            return ReactorNettyClient.connect(configuration.toClientConfiguration(), configuration.getApplicationName(), configuration.getConnectionId()).cast(Client.class);
-        }), configuration);
+        this(MssqlConnectionFactory::connect, configuration);
     }
 
-    MssqlConnectionFactory(Mono<? extends Client> clientFactory, MssqlConnectionConfiguration configuration) {
+    MssqlConnectionFactory(Function<MssqlConnectionConfiguration, Mono<? extends Client>> clientFactory,
+                           MssqlConnectionConfiguration configuration) {
+
         this.clientFactory = Assert.requireNonNull(clientFactory, "clientFactory must not be null");
         this.configuration = Assert.requireNonNull(configuration, "configuration must not be null");
         this.connectionOptions = configuration.toConnectionOptions();
     }
 
-    @Override
-    public Mono<MssqlConnection> create() {
+    private static Mono<? extends Client> connect(MssqlConnectionConfiguration configuration) {
+
+        return Mono.defer(() -> {
+            Assert.requireNonNull(configuration, "configuration must not be null");
 
-        LoginConfiguration loginConfiguration = this.configuration.getLoginConfiguration();
+            return ReactorNettyClient.connect(configuration.toClientConfiguration(), configuration.getApplicationName(),
+                configuration.getConnectionId());
+        });
+    }
 
-        return this.clientFactory.delayUntil(client -> {
+    private Mono<? extends Client> initializeClient(final MssqlConnectionConfiguration configuration, boolean allowReroute) {
+
+        LoginConfiguration loginConfiguration = configuration.getLoginConfiguration();
+
+        return this.clientFactory.apply(configuration).flatMap(client -> {
             return LoginFlow.exchange(client, loginConfiguration)
-                .doOnError(e -> client.close().subscribe());
-        })
+                .doOnError(e -> client.close().subscribe())
+                .flatMap(result -> {
+                    switch (result.getOutcome()) {
+
+                        case CONNECTED:
+                            return Mono.just(client);
+
+                        case ROUTED: {
+                            if (!allowReroute) {
+                                return Mono.error(
+                                    new IllegalStateException("Client was redirected more than once."));
+                            }
+
+                            assert result.getAlternateServerName() != null;
+
+                            MssqlConnectionConfiguration routeConfiguration = configuration.toBuilder()
```

Comment:
Instead using a builder, we could simply have a `withServer(…)` method that creates a new `MssqlConnectionConfiguration` object. Since both types are in the same package, we don't need to expose this method publicly.

Reply:
Done. I have left the `toBuilder` as a private method for now. Let me know if you prefer to have it gone altogether.

Reply:
There's no need for a `toBuilder` method, so let's remove it. These kind of methods are error prone for adding additional properties as they get easily forgotten. Extending a constructor typically causes compile errors if a caller wasn't updated.